### PR TITLE
Fix flexvolume Cleanup

### PR DESCRIPTION
**Pull request by vefimova**
- fixed recursive call during secret removal (actually we never call it due to the below fixed issue)
- moved volume cleanup to the end of StopDomain call as by the time of RemoveContainer all flexvolume data is already cleaned by kubelet's VolumeManager

---
This change is [Reviewable](https://reviewable.io/reviews/mirantis/virtlet/336)

### CODE REVIEW

File: pkg/libvirttools/virtualization.go [hunk 1]
```diff
@@ -406,6 +408,7 @@ func (v *VirtualizationTool) startContainer(containerId string) error {
 
 	state, err := domain.State()
 	if err != nil {
+<<<<<<< d36335959d762c89cc6ad920f4eda6aaeb6e678b
```
Comment (jellonek) [inline pkg/libvirttools/virtualization.go]: Please remove this from commit (probably by amending last commit to this).

File: examples/util.sh [hunk 2]
```diff
@@ -0,0 +1,22 @@
+#!/bin/bash
```
Comment (jellonek) [inline examples/util.sh]: Typo in commit message - cleanup.

**Comment by ivan4th** [comment]
Review status: 0 of 19 files reviewed at latest revision, 8 unresolved discussions.

---

*[examples/util.sh, line 22 at r1](https://reviewable.io:443/reviews/mirantis/virtlet/336#-KnEUqp9IUOZQgG2YsO3:-KnEUqp9IUOZQgG2YsO4:bpcjtcw) ([raw file](https://github.com/mirantis/virtlet/blob/64c901eb247d18e168eae97df2b9fccb6ede67b5/examples/util.sh#L22)):*
> ```Shell
>   kubectl get pod ${namespace_opts} "${pod_name}" \
>           -o 'jsonpath={.status.containerStatuses[0].containerID}-{.status.containerStatuses[0].name}'|sed 's/.*__//'
> }
> ```

I'm not sure we should do this. Problem is, virsh.sh is supposed to be "self-contained", as in, some docs suggest downloading and running it. We need to either update all the docs or just not source anything from virsh.sh

---

*[pkg/libvirttools/ceph_flexvolume.go, line 139 at r1](https://reviewable.io:443/reviews/mirantis/virtlet/336#-KnE8t0_LFxMt-D0b3Md:-KnE8t0_LFxMt-D0b3Me:bxfho3x) ([raw file](https://github.com/mirantis/virtlet/blob/64c901eb247d18e168eae97df2b9fccb6ede67b5/pkg/libvirttools/ceph_flexvolume.go#L139)):*
> ```Go
> 	case err == virt.ErrSecretNotFound:
> 		// ok, no need to delete the secret
>                glog.V(3).Infof("No secret with uuid %q for ceph volume was found.", v.secretUuid())
> ```

Please remove the dot after "found"

---

*[pkg/libvirttools/flexvolume_volumesource.go, line 45 at r1](https://reviewable.io:443/reviews/mirantis/virtlet/336#-KnEVUCX60T2mhN6BIqy:-KnEVUCX60T2mhN6BIqz:b-ikhcge) ([raw file](https://github.com/mirantis/virtlet/blob/64c901eb247d18e168eae97df2b9fccb6ede67b5/pkg/libvirttools/flexvolume_volumesource.go#L45)):*
> ```Go
> 	dir := filepath.Join(owner.KubeletRootDir(), config.PodSandboxId, flexvolumeSubdir)
> 	if _, err := os.Stat(dir); os.IsNotExist(err) {
> 		glog.V(2).Infof("No flexvolumes to process for %q with uuid %q.", config.Name, config.DomainUUID)
> ```

Perhaps we don't need the dot after `%q`

---

*[pkg/libvirttools/libvirt_domain.go, line 176 at r1](https://reviewable.io:443/reviews/mirantis/virtlet/336#-KnEVEkZ6X3jlGVvT-S3:-KnEVEkZ6X3jlGVvT-S4:b-oemhsm) ([raw file](https://github.com/mirantis/virtlet/blob/64c901eb247d18e168eae97df2b9fccb6ede67b5/pkg/libvirttools/libvirt_domain.go#L176)):*
> ```Go
> 
> func (secret *LibvirtSecret) Remove() error {
> 	return secret.s.Undefine()
> ```

Good catch! :+1: 

---

*[pkg/libvirttools/TestDomainDefinitions__ceph_flexvolume.json, line 494 at r1](https://reviewable.io:443/reviews/mirantis/virtlet/336#-KnEW0zmiX6AlOAL3Um3:-KnEW0znjtVbnH-zcf_2:b-h863zd) ([raw file](https://github.com/mirantis/virtlet/blob/64c901eb247d18e168eae97df2b9fccb6ede67b5/pkg/libvirttools/TestDomainDefinitions__ceph_flexvolume.json#L494)):*
> ```JSON
>   },
>   {
>     "name": "domain conn: 231700d5-c9a6-5a49-738d-99a954c51550-container1: Shutdown"
> ```

Why are we shutting down the domain that was not started?

---

*[pkg/libvirttools/virtualization.go, line 571 at r1](https://reviewable.io:443/reviews/mirantis/virtlet/336#-KnEWNYic6ex7KeFiv20:-KnEWNYjbDyy6acjEdDN:b1ue22f) ([raw file](https://github.com/mirantis/virtlet/blob/64c901eb247d18e168eae97df2b9fccb6ede67b5/pkg/libvirttools/virtualization.go#L571)):*
> ```Go
> }
> 
> func (v *VirtualizationTool) volumeCleanup(containerId string) error {
> ```

maybe `cleanupVolumes`?

---

*[pkg/libvirttools/virtualization.go, line 647 at r1](https://reviewable.io:443/reviews/mirantis/virtlet/336#-KnEWrT5Mg1S_Nfwn4ZP:-KnEWrT5Mg1S_Nfwn4ZQ:b-2ghp41) ([raw file](https://github.com/mirantis/virtlet/blob/64c901eb247d18e168eae97df2b9fccb6ede67b5/pkg/libvirttools/virtualization.go#L647)):*
> ```Go
> 
> 	if config == nil {
> 		glog.Warningf("No info found for domain %q in metadata store. Domain cleanup skipped.", containerId)
> ```

Please remove the dot at the end of the message

---

*[tests/e2e/e2e.sh, line 59 at r1](https://reviewable.io:443/reviews/mirantis/virtlet/336#-KnEX5zgbqd_Jzv_pb1G:-KnEX5zgbqd_Jzv_pb1H:bn15g17) ([raw file](https://github.com/mirantis/virtlet/blob/64c901eb247d18e168eae97df2b9fccb6ede67b5/tests/e2e/e2e.sh#L59)):*
> ```Shell
>   # for some reason even after 'kubectl get pod' check returns 'NOT FOUND'
>   # domain remains defined being in "shut off" state for noticable time
>   sleep 60
> ```

We need to check what's going on here. This may mean that the cleanup mechanism is not doing its job. I'd suggest the following debugging scheme: 1) edit `deploy/virtlet-ds-dev.yaml` to set `VIRTLET_LOGLEVEL=1` 2) Add `glog.V(0).Infof(...)` to critical places like undefining domains etc., you can do this in `pkg/libvirttools/libvirt_domain.go` etc 3) Switch some critical CRI calls in `pkg/manager/manager.go` to V(0) logging 4) Reproduce this situation and look at the sequence of CRI calls during this "noticeable time"

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/mirantis/virtlet/336)*

**Comment by vefimova** [comment]
Review status: 0 of 16 files reviewed at latest revision, 7 unresolved discussions.

---

*[pkg/libvirttools/ceph_flexvolume.go, line 139 at r1](https://reviewable.io:443/reviews/mirantis/virtlet/336#-KnE8t0_LFxMt-D0b3Md:-KnKl7dDPOPsUT9y14CG:b-896fix) ([raw file](https://github.com/mirantis/virtlet/blob/64c901eb247d18e168eae97df2b9fccb6ede67b5/pkg/libvirttools/ceph_flexvolume.go#L139)):*
<details><summary><i>Previously, ivan4th (Ivan Shvedunov) wrote…</i></summary><blockquote>

Please remove the dot after "found"
</blockquote></details>

Done.

---

*[pkg/libvirttools/flexvolume_volumesource.go, line 45 at r1](https://reviewable.io:443/reviews/mirantis/virtlet/336#-KnEVUCX60T2mhN6BIqy:-KnKlGG9PO3LUlyX-6uJ:b-896fix) ([raw file](https://github.com/mirantis/virtlet/blob/64c901eb247d18e168eae97df2b9fccb6ede67b5/pkg/libvirttools/flexvolume_volumesource.go#L45)):*
<details><summary><i>Previously, ivan4th (Ivan Shvedunov) wrote…</i></summary><blockquote>

Perhaps we don't need the dot after `%q`
</blockquote></details>

Done.

---

*[pkg/libvirttools/virtualization.go, line 571 at r1](https://reviewable.io:443/reviews/mirantis/virtlet/336#-KnEWNYic6ex7KeFiv20:-KnL00N-Q7B97Bs2BbMX:b-896fix) ([raw file](https://github.com/mirantis/virtlet/blob/64c901eb247d18e168eae97df2b9fccb6ede67b5/pkg/libvirttools/virtualization.go#L571)):*
<details><summary><i>Previously, ivan4th (Ivan Shvedunov) wrote…</i></summary><blockquote>

maybe `cleanupVolumes`?
</blockquote></details>

Done.

---

*[pkg/libvirttools/virtualization.go, line 647 at r1](https://reviewable.io:443/reviews/mirantis/virtlet/336#-KnEWrT5Mg1S_Nfwn4ZP:-KnKmqfPOXN_DilI-R52:b-896fix) ([raw file](https://github.com/mirantis/virtlet/blob/64c901eb247d18e168eae97df2b9fccb6ede67b5/pkg/libvirttools/virtualization.go#L647)):*
<details><summary><i>Previously, ivan4th (Ivan Shvedunov) wrote…</i></summary><blockquote>

Please remove the dot at the end of the message
</blockquote></details>

Done.

---

*[tests/e2e/e2e.sh, line 59 at r1](https://reviewable.io:443/reviews/mirantis/virtlet/336#-KnEX5zgbqd_Jzv_pb1G:-KnLaaioCkFvxXlY-Cwf:b-tkos0l) ([raw file](https://github.com/mirantis/virtlet/blob/64c901eb247d18e168eae97df2b9fccb6ede67b5/tests/e2e/e2e.sh#L59)):*
<details><summary><i>Previously, ivan4th (Ivan Shvedunov) wrote…</i></summary><blockquote>

We need to check what's going on here. This may mean that the cleanup mechanism is not doing its job. I'd suggest the following debugging scheme: 1) edit `deploy/virtlet-ds-dev.yaml` to set `VIRTLET_LOGLEVEL=1` 2) Add `glog.V(0).Infof(...)` to critical places like undefining domains etc., you can do this in `pkg/libvirttools/libvirt_domain.go` etc 3) Switch some critical CRI calls in `pkg/manager/manager.go` to V(0) logging 4) Reproduce this situation and look at the sequence of CRI calls during this "noticeable time"
</blockquote></details>

Filed an issue https://github.com/Mirantis/virtlet/issues/341

---

*[examples/util.sh, line 22 at r1](https://reviewable.io:443/reviews/mirantis/virtlet/336#-KnEUqp9IUOZQgG2YsO3:-KnLadoh8I9K4sATyEh-:b-896fix) ([raw file](https://github.com/mirantis/virtlet/blob/64c901eb247d18e168eae97df2b9fccb6ede67b5/examples/util.sh#L22)):*
<details><summary><i>Previously, ivan4th (Ivan Shvedunov) wrote…</i></summary><blockquote>

I'm not sure we should do this. Problem is, virsh.sh is supposed to be "self-contained", as in, some docs suggest downloading and running it. We need to either update all the docs or just not source anything from virsh.sh
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/mirantis/virtlet/336)*

File: pkg/libvirttools/virtualization.go (hunk 1 above)
Comment (vefimova) [inline pkg/libvirttools/virtualization.go]: done

File: examples/util.sh (hunk 2 above)
Comment (vefimova) [inline examples/util.sh]: done

**Comment by ivan4th** [comment]
Review status: 0 of 16 files reviewed at latest revision, 8 unresolved discussions.

---

*[pkg/libvirttools/TestDomainDefinitions__raw_devices.json, line 461 at r2](https://reviewable.io:443/reviews/mirantis/virtlet/336#-Knf7h-dPbnVRhy17xY1:-Knf7h-dPbnVRhy17xY2:blbl64a) ([raw file](https://github.com/mirantis/virtlet/blob/603da982a8b15612956a925d331feb04bfe4da1d/pkg/libvirttools/TestDomainDefinitions__raw_devices.json#L461)):*
> ```JSON
>   {
>     "name": "storage: volumes: RemoveVolumeByName",
>     "data": "root_231700d5-c9a6-5a49-738d-99a954c51550"
> ```

This test differs from e.g. `TestContainerLifecycle` in that it doesn't stop the container before removing them. So `RemoveContainer` is not cleaning up the volumes (and maybe not removing the domain, thus this 60sec delay in the test?) This is incorrect behavior -- `RemoveContainer` should stop the container correctly (although forcibly) if it's running https://github.com/kubernetes/kubernetes/blob/v1.6.6/pkg/kubelet/api/v1alpha1/runtime/api.proto#L53 So leaving volumes here is incorrect.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/mirantis/virtlet/336)*

**Comment by vefimova** [comment]
Review status: 0 of 16 files reviewed at latest revision, 8 unresolved discussions.

---

*[pkg/libvirttools/TestDomainDefinitions__raw_devices.json, line 461 at r2](https://reviewable.io:443/reviews/mirantis/virtlet/336#-Knf7h-dPbnVRhy17xY1:-KnfD9g8wSOR6BQ0rUjB:bfh2n25) ([raw file](https://github.com/mirantis/virtlet/blob/603da982a8b15612956a925d331feb04bfe4da1d/pkg/libvirttools/TestDomainDefinitions__raw_devices.json#L461)):*
<details><summary><i>Previously, ivan4th (Ivan Shvedunov) wrote…</i></summary><blockquote>

This test differs from e.g. `TestContainerLifecycle` in that it doesn't stop the container before removing them. So `RemoveContainer` is not cleaning up the volumes (and maybe not removing the domain, thus this 60sec delay in the test?) This is incorrect behavior -- `RemoveContainer` should stop the container correctly (although forcibly) if it's running https://github.com/kubernetes/kubernetes/blob/v1.6.6/pkg/kubelet/api/v1alpha1/runtime/api.proto#L53 So leaving volumes here is incorrect.
</blockquote></details>

so your previous comment was "Why are we shutting down the domain that was not started?", so as tests have "TestDomainDefinitions" prefix, I decided to remove added change of calling StopContainer, or you think it's better to add "StartContainer" otherwise to test all sequence for these "TestDomainDefinitions"?

The fix of cleanup was to move volume cleanup from `RemoveContainer` to `StopContainer` as by the time of  `RemoveContainer` all flexvolume dirs with data are already cleaned by kubelet's  volume manager. If stop container retruned failure, i thought that kubelet will keep trying and won't call RemoveContainer in other way am I wrong here?
All other cleanup actions on `RemoveContainer` wasn't touched (see https://github.com/Mirantis/virtlet/pull/336/files#diff-27a11a3eecdfb4f023d0be5eb326dee3R592).

re 60sec delay in the test - if you're referring to  https://github.com/Mirantis/virtlet/issues/341 - domain disapears eventually (it won't happen without `undefine` call), and all volumes and secrets are cleared as well. I suspect libvirt `undefine` call should also be waited by us as `shutdown` and `destroy`.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/mirantis/virtlet/336)*

**Comment by jellonek** [comment]
Reviewed 2 of 19 files at r1, 4 of 17 files at r2, 5 of 11 files at r3.
Review status: all files reviewed at latest revision, 9 unresolved discussions.

---

*[pkg/libvirttools/virtualization.go, line 447 at r3](https://reviewable.io:443/reviews/mirantis/virtlet/336#-KnnAeEmVtJRDIcC-NCh:-KnnAeEmVtJRDIcC-NCi:b-1z0hd0) ([raw file](https://github.com/mirantis/virtlet/blob/95263b7bad28b765200eb5db3b10bb5072bf8f7a/pkg/libvirttools/virtualization.go#L447)):*
> ```Go
> 	}
> ```

Why there is additional empty line?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/mirantis/virtlet/336)*

**Comment by vefimova** [comment]
Review status: 7 of 8 files reviewed at latest revision, 9 unresolved discussions.

---

*[pkg/libvirttools/virtualization.go, line 447 at r3](https://reviewable.io:443/reviews/mirantis/virtlet/336#-KnnAeEmVtJRDIcC-NCh:-KnnDk7-lncTu0myS1g6:b-hfpv0l) ([raw file](https://github.com/mirantis/virtlet/blob/95263b7bad28b765200eb5db3b10bb5072bf8f7a/pkg/libvirttools/virtualization.go#L447)):*
<details><summary><i>Previously, jellonek (Piotr Skamruk) wrote…</i></summary><blockquote>

Why there is additional empty line?
</blockquote></details>

removed

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/mirantis/virtlet/336)*

**Comment by jellonek** [comment]
:lgtm:

---

Reviewed 1 of 1 files at r4.
Review status: all files reviewed at latest revision, 8 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/mirantis/virtlet/336#-:-KnnDt9g-K_oGMYVGSg_:bnfp4nl)*

**Comment by ivan4th** [comment]
Review status: all files reviewed at latest revision, 2 unresolved discussions.

---

*[tests/e2e/e2e.sh, line 59 at r1](https://reviewable.io:443/reviews/mirantis/virtlet/336#-KnEX5zgbqd_Jzv_pb1G:-KnoCqVk_9ciQFCl0vkP:b2nhll9) ([raw file](https://github.com/mirantis/virtlet/blob/64c901eb247d18e168eae97df2b9fccb6ede67b5/tests/e2e/e2e.sh#L59)):*
<details><summary><i>Previously, vefimova (Victoria Efimova) wrote…</i></summary><blockquote>

Filed an issue https://github.com/Mirantis/virtlet/issues/341
</blockquote></details>

Ok.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/mirantis/virtlet/336)*

**Comment by ivan4th** [comment]
:lgtm:

---

Review status: all files reviewed at latest revision, 1 unresolved discussion.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/mirantis/virtlet/336#-:-KnoD-gIlp0XvMpdo8eI:bnfp4nl)*

**Pull request merged**